src/nnet3bin/nnet3-discriminative-compute-from-egs.cc

```diff
@@ -0,0 +1,149 @@
+// nnet3bin/nnet3-discriminative-compute-from-egs.cc
+
+// Copyright 2015  Johns Hopkins University (author: Daniel Povey)
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include "base/kaldi-common.h"
+#include "util/common-utils.h"
+#include "hmm/transition-model.h"
+#include "nnet3/nnet-nnet.h"
+#include "nnet3/nnet-example-utils.h"
+#include "nnet3/nnet-discriminative-example.h"
+#include "nnet3/nnet-optimize.h"
+
+
+namespace kaldi {
+namespace nnet3 {
+
+class NnetComputerFromEg {
+ public:
+  NnetComputerFromEg(const Nnet &nnet):
+      nnet_(nnet), compiler_(nnet) { }
+
+  // Compute the output (which will have the same number of rows as the number
+  // of Indexes in the output of the eg), and put it in "output".
+  void Compute(const NnetExample &eg, Matrix<BaseFloat> *output) {
+    ComputationRequest request;
+    bool need_backprop = false, store_stats = false;
+    GetComputationRequest(nnet_, eg, need_backprop, store_stats, &request);
+    const NnetComputation &computation = *(compiler_.Compile(request));
+    NnetComputeOptions options;
+    if (GetVerboseLevel() >= 3)
+      options.debug = true;
+    NnetComputer computer(options, computation, nnet_, NULL);
+    computer.AcceptInputs(nnet_, eg.io);
+    computer.Forward();
+    const CuMatrixBase<BaseFloat> &nnet_output = computer.GetOutput("output");
+    output->Resize(nnet_output.NumRows(), nnet_output.NumCols());
+    nnet_output.CopyToMat(output);
+  }
+ private:
+  const Nnet &nnet_;
+  CachingOptimizingCompiler compiler_;
+  
+};
+
+}
+}
+
+int main(int argc, char *argv[]) {
+  try {
+    using namespace kaldi;
+    using namespace kaldi::nnet3;
+    typedef kaldi::int32 int32;
+    typedef kaldi::int64 int64;
+
+    const char *usage =
+        "Read input nnet discriminative training examples, and compute the "
+        "output for each one. This program is similar to "
+        "nnet3-compute-from-egs, but works with discriminative egs. \n"
+        "If --apply-exp=true, apply the Exp() function to the output before writing\n"
```

Comment:
you should clarify in the usage message that it only uses the features in the discriminative egs, it does not do any forward-backward on the lattice.